src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs

```diff
@@ -150,7 +150,15 @@ public static bool IsLockFileStillValid(DependencyGraphSpec dgSpec, PackagesLock
 
         private static bool HasProjectDependencyChanged(IEnumerable<LibraryDependency> newDependencies, IEnumerable<LockFileDependency> lockFileDependencies)
         {
-            foreach (var dependency in newDependencies.Where(dep => dep.LibraryRange.TypeConstraint == LibraryDependencyTarget.Package))
+            // If the count is not the same, something has changed.
+            // Otherwise we N^2 walk below determines whether anything has changed.
+            var newPackageDependencies = newDependencies.Where(dep => dep.LibraryRange.TypeConstraint == LibraryDependencyTarget.Package);
+            if(newPackageDependencies.Count() != lockFileDependencies.Count())
+            {
+                return true;
+            }
+
+            foreach (var dependency in newPackageDependencies)
```

Comment:
So while this works in this case, it's not generally safe to enumerate an IEnumerable more than once (`newPackageDependencies.Count()` was the first). Think of an IEnumerable that comes from a database query result set (ok, not going to happen in the NuGet client code), or streaming HTTP response, or possibly methods that use `yield return`. I'd feel more comfortable if we didn't enumerate multiple times, as that will avoid bugs if the value of newDependencies happens to ever be a type that can't be re-enumerated in the future.

Reply:
That's definitely a valid concern :)
I did this consciously.
I'd rather enumerate twice than create a new list by calling ToList().